File: scripts/serve/vitals.py
"""The vitals store: continuous stream health as a time series.

WHAT THIS IS FOR. Video and audio are CONTINUOUS. "Did the stream work" is not
a question about a call or an event, so neither the trace lane nor the log lane
answers it, and `analytics.db` cannot either — its counters have no per-sample
timestamp, only a day bucket, which is honest for "how many restores yesterday"
and a lie at any finer resolution. This store holds timestamped samples at
sub-minute resolution, and it is the thing that survives dropping Instana: an
OTLP export is a CONSUMER of it (`vitals_otlp.py`), never its purpose.

WHAT IT IS NOT FOR. Discrete degradations — a stall, a decode error, an ABR
downshift, blocked audio — are EVENTS, and they belong on the log and event
planes. What lives here is the CONTINUOUS number a threshold on such an event
would fire against, which is why `frames_dropped` is a cumulative counter here
and "the stream stalled" is not a row here at all.

POSTURE, inherited wholesale from the log lane one file over. INGEST IS OPEN
(`POST /vitals`, origin-checked on the public listener): a tab whose stream is
bad has to be able to say so without holding an admin session, and that is the
tab whose numbers matter most. READS ARE ADMIN-ONLY (`/auth/vitals/*`,
`vitals_read.py`): a sample names a station and a session.

RETENTION IS 30 DAYS (`RETENTION_DAYS`), the LONGEST window in the plane, and
that is deliberate rather than careless. The instinct with sub-minute data is
to keep a tight window because it is dense — but density is a reason to size
the DISK, not to throw the data away, and this box has ~166 GB free against a
measured worst case of a few hundred MB for the whole window (the arithmetic is
in `docs/ANALYTICS.md` §8.6). A month means "has this station always been like
this, or did it change?" is answerable, which is the question a one-visitor
gallery actually asks; three days would have answered only "is it bad now",
which the live read already answers for free.

The RUNAWAY BACKSTOP below is what handles the case retention used to: a
producer stuck in a flush loop is a fault, and a fault gets a ceiling. A
budget is not the same thing as a ceiling and should not be spelt as one.

There is NO DOWNSAMPLING. A rollup is a second schema, a second prune and a
second thing to be wrong; at this volume the full-resolution window is small
enough that buying one would be paying for a problem we do not have. If the
gallery ever has real traffic, the honest first move is a bigger disk and the
second is a rollup — in that order.
"""

from __future__ import annotations

import json
import re
import sqlite3
import threading
import time
from pathlib import Path

import vitals_schema
from vitals_schema import COLUMNS

IDENT_RE = re.compile(r"^[A-Za-z0-9._@/+-]{1,64}$")

#: Samples one batch may carry. A client sampling at 1 Hz and flushing on a 20 s
#: timer offers 20; a pagehide flush after a backgrounded stretch offers up to
#: its whole 600-deep queue. 1024 clears that with room and still bounds a
#: runaway producer.
MAX_SAMPLES_PER_BATCH = 1024
BODY_MAX = 512 * 1024
RETENTION_DAYS = 30
#: Runaway backstop, checked on prune. NOT a capacity budget — a ceiling on a
#: FAULT. Legitimate saturation (every one of the 71 stations streaming
#: continuously for the whole 30-day window at 1 Hz) would be 184 M rows and is
#: not a thing that can happen in a gallery with one visitor; a client stuck in
#: a flush loop can produce arbitrarily many in an afternoon, and that is what
#: this stops. Set well above any plausible real load precisely so that hitting
#: it is diagnostic rather than routine.
MAX_ROWS = 20_000_000
#: How long a shipped batch's id is remembered so a re-ship stores nothing.
#: Same mechanism, same reason, as `logs.BATCH_MEMORY_DAYS`.
BATCH_MEMORY_DAYS = 1
#: Sane bounds for a stored value. A vital is a physical measurement, so an
#: infinity or a 1e300 is a producer bug, and sqlite would store it happily and
#: hand it to a chart that then has no y-axis. Clamped rather than refused: the
#: rest of the sample is still evidence.
VALUE_MIN, VALUE_MAX = -1e9, 1e9


def _day(ts_ms: int) -> str:
    return time.strftime("%Y-%m-%d", time.gmtime(ts_ms / 1000))


def _ident(v, default: str = "unknown") -> str:
    return v if isinstance(v, str) and IDENT_RE.match(v) else default


def _num(v):
    """A finite float, or None. Booleans are NOT numbers here: `True` for a
    gauge would store 1.0 and read as a measurement rather than as a flag the
    producer forgot to convert."""
    if isinstance(v, bool) or not isinstance(v, (int, float)):
        return None
    f = float(v)
    if f != f or f in (float("inf"), float("-inf")):
        return None
    return max(VALUE_MIN, min(VALUE_MAX, f))


class VitalsStore:
    """Timestamped stream-health samples. Safe from any thread."""

    def __init__(self, path: Path, read_only: bool = False):
        """`read_only` is for a REPORT or for the forwarder — same seam, same
        reason, as `traces.TraceStore` and `logs.LogStore`: opening the live
        store the normal way runs DDL and a migration against the file the
        serving plane is writing, and a script whose whole job is to read a
        number should take none of that risk."""
        self.path = Path(path)
        self._lock = threading.RLock()
        if read_only:
            self._db = sqlite3.connect(f"file:{self.path}?mode=ro", uri=True, check_same_thread=False)
            return
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._db = sqlite3.connect(str(self.path), check_same_thread=False)
        self._db.execute("PRAGMA journal_mode=WAL")
        self._db.executescript(vitals_schema.SCHEMA)
        vitals_schema.migrate(self._db)
        self._db.commit()

    # ---- intake --------------------------------------------------------

    def record(self, batch: dict) -> int:
        """Fold one producer's samples in. Returns how many were accepted.

        The envelope mirrors `POST /logs` and `POST /traces` — a `resource`
        naming the producer once, and a list whose entries carry only what
        varies — because a third envelope shape for a third pillar is three
        parsers to keep in step for no gain:

          {"resource": {"service.instance.id": <station>, "session.id": ...,
                        "kh.bundle": ..., "kh.source": "spa"|"daemon"},
           "batchId": "...",
           "samples": [{"t": <ms>, "v": {"fps": 30, "rtt_ms": 12.4, ...}}]}

        A key in `v` that is not in the catalogue is IGNORED rather than
        refused. A client that shipped early, or a build that knows a vital
        this deploy does not, still lands everything it and we agree on — the
        alternative loses the whole sample to punish one field.
        """
        res = batch.get("resource") if isinstance(batch.get("resource"), dict) else {}
        station = _ident(res.get("service.instance.id"))
        session = _ident(res.get("session.id"))
        build = _ident(res.get("kh.bundle"))
        source = _ident(res.get("kh.source"), "spa")
        raw = batch.get("samples")
        if not isinstance(raw, list):
            return 0
        observed = int(time.time() * 1000)
        # Claimed BEFORE any row is built, so two concurrent ships of the same
        # spool file cannot both pass: the PRIMARY KEY is the lock, never a
        # preceding SELECT. Straight out of `logs.LogStore.record`.
        if not self._claim_batch(batch.get("batchId"), observed):
            return 0
        rows = []
        for rec in raw[:MAX_SAMPLES_PER_BATCH]:
            if not isinstance(rec, dict):
                continue
            vals = rec.get("v")
            if not isinstance(vals, dict):
                continue
            cells = [_num(vals.get(c)) for c in COLUMNS]
            # A sample with no readable value at all is not a sample. Storing
            # it would put a row on every chart's x-axis carrying nothing,
            # which reads as "we measured, and it was nothing".
            if all(c is None for c in cells):
                continue
            ts = rec.get("t")
            ts = int(ts) if isinstance(ts, (int, float)) and not isinstance(ts, bool) and ts > 0 else observed
            rows.append(
                (
                    ts,
                    observed,
                    _ident(rec.get("st"), station),
                    _ident(rec.get("sid"), session),
                    source,
                    build,
                    _day(ts),
                    *cells,
                )
            )
        if not rows:
            return 0
        cols = "ts_ms,observed_ms,station,session_id,source,build,day," + ",".join(COLUMNS)
        marks = ",".join("?" * (7 + len(COLUMNS)))
        with self._lock:
            self._db.executemany(f"INSERT INTO vital({cols}) VALUES({marks})", rows)
            self._db.commit()
        return len(rows)

    def _claim_batch(self, batch_id, observed: int) -> bool:
        """True when this batch has not been seen. An unnamed batch always
        passes — idempotency is a property a producer opts into by naming
        itself, and the browser's live sink genuinely has nothing to dedupe."""
        if not isinstance(batch_id, str) or not batch_id:
            return True
        with self._lock:
            cur = self._db.cursor()
            cur.execute("INSERT OR IGNORE INTO vital_batch(batch_id,seen_ms) VALUES(?,?)", (batch_id[:128], observed))
            self._db.commit()
            return cur.rowcount > 0

    # ---- reads ---------------------------------------------------------

    def series(self, **f) -> dict:
        """A page of samples, oldest first — which is the order a CHART wants,
        and the opposite of what the log and trace lanes default to. Those are
        read as "the most recent N"; a time series is read as a line, and a
        reader that has to reverse every page before plotting it will one day
        forget to.

        `order="ingest"` walks by `seq` instead, which is what the forwarder's
        watermark needs.
        """
        where, args = [], []
        for col, key in (("station", "station"), ("session_id", "session"), ("source", "source"), ("build", "build")):
            if f.get(key):
                where.append(f"{col}=?")
                args.append(str(f[key])[:64])
        for col, key, op in (("ts_ms", "since_ms", ">="), ("ts_ms", "until_ms", "<="), ("seq", "since_seq", ">")):
            if f.get(key) is not None:
                where.append(f"{col}{op}?")
                args.append(int(f[key]))
        sql = " WHERE " + " AND ".join(where) if where else ""
        limit = max(1, min(int(f.get("limit") or 500), 5000))
        offset = max(0, int(f.get("offset") or 0))
        order = "seq ASC" if f.get("order") == "ingest" else "ts_ms ASC, seq ASC"
        cols = "seq,ts_ms,observed_ms,station,session_id,source,build," + ",".join(COLUMNS)
        with self._lock:
            cur = self._db.cursor()
            total = cur.execute(f"SELECT COUNT(*) FROM vital{sql}", args).fetchone()[0]
            rows = cur.execute(
                f"SELECT {cols} FROM vital{sql} ORDER BY {order} LIMIT ? OFFSET ?", (*args, limit, offset)
            ).fetchall()
        return {"samples": [self._row(r) for r in rows], "total": total, "limit": limit, "offset": offset}

    @staticmethod
    def _row(r) -> dict:
        """One stored row as JSON. NULL VITALS ARE OMITTED, not sent as null:
        most of every row is the other producer's columns, and a wire that
        spelled all of them out would be mostly the word "null"."""
        return {
            "seq": r[0],
            "tsMs": r[1],
            "observedMs": r[2],
            "station": r[3],
            "sessionId": r[4],
            "source": r[5],
            "build": r[6],
            "v": {c: r[7 + i] for i, c in enumerate(COLUMNS) if r[7 + i] is not None},
        }

    def live(self, within_ms: int = 120_000) -> dict:
        """THE OPERATOR'S FIRST READ: the newest sample for every station that
        has produced one recently, one row each. It answers "is anything
        streaming right now, and is it healthy" without paging a series per
        station, and it is the read the /admin surface opens with.

        `within_ms` is what "recently" means. Default two minutes: long enough
        that a client on a 60 s flush timer is never missed, short enough that
        a station nobody is on drops off the list rather than showing a stale
        line as if it were current.
        """
        cutoff = int(time.time() * 1000) - max(1000, within_ms)
        cols = "seq,ts_ms,observed_ms,station,session_id,source,build," + ",".join(COLUMNS)
        with self._lock:
            rows = self._db.execute(
                f"SELECT {cols} FROM vital WHERE seq IN "
                "(SELECT MAX(seq) FROM vital WHERE ts_ms>=? GROUP BY station,session_id,source) "
                "ORDER BY station,session_id",
                (cutoff,),
            ).fetchall()
        return {"live": [self._row(r) for r in rows], "withinMs": within_ms}

    def facets(self, since_ms: int) -> dict:
        """What is in the window, and — the number this lane is judged on — how
        much wall-clock time is actually COVERED by samples. A vitals store
        that holds a thousand rows from one 40-second session is not
        monitoring; it is a souvenir."""
        with self._lock:
            cur = self._db.cursor()

            def group(col):
                return [
                    {"value": v, "n": n}
                    for v, n in cur.execute(
                        f"SELECT {col},COUNT(*) FROM vital WHERE ts_ms>=? "
                        f"GROUP BY {col} ORDER BY COUNT(*) DESC LIMIT 100",
                        (since_ms,),
                    )
                ]

            span = cur.execute(
                "SELECT COUNT(*),MIN(ts_ms),MAX(ts_ms),COUNT(DISTINCT session_id) FROM vital WHERE ts_ms>=?",
                (since_ms,),
            ).fetchone()
        return {
            "stations": group("station"),
            "sources": group("source"),
            "total": span[0] or 0,
            "firstMs": span[1],
            "lastMs": span[2],
            "sessions": span[3] or 0,
            "catalogue": [{"column": c, "metric": m, "unit": u, "kind": k} for c, m, u, k in vitals_schema.CATALOGUE],
        }

    def prune(self, keep_days: int = RETENTION_DAYS, max_rows: int = MAX_ROWS) -> int:
        """Drop samples older than the window; then, only if still over the
        runaway backstop, drop the oldest regardless of age. A sample is whole
        on its own, like a log record and unlike a trace, so the size backstop
        can be a plain oldest-first delete."""
        cutoff = _day(int((time.time() - keep_days * 86400) * 1000))
        with self._lock:
            cur = self._db.cursor()
            cur.execute("DELETE FROM vital WHERE day<?", (cutoff,))
            dropped = cur.rowcount
            cur.execute(
                "DELETE FROM vital_batch WHERE seen_ms<?", (int((time.time() - BATCH_MEMORY_DAYS * 86400) * 1000),)
            )
            over = cur.execute("SELECT COUNT(*) FROM vital").fetchone()[0] - max_rows
            if over > 0:
                cur.execute("DELETE FROM vital WHERE seq IN (SELECT seq FROM vital ORDER BY seq LIMIT ?)", (over,))
                dropped += cur.rowcount
            self._db.commit()
        return dropped

    def close(self) -> None:
        with self._lock:
            self._db.close()


def export_json(found: dict) -> str:
    return json.dumps(found, separators=(",", ":"))

File: spa/src/three/streamClient/vitals.ts
// ============================================================================
//  streamClient/vitals — continuous stream health as a TIME SERIES.
//  ---------------------------------------------------------------------------
//  WHY THIS EXISTS. Every number below was already being computed, once every
//  five seconds, and rendered into a 140-character string that went to
//  `clientlog.jsonl` — a sample, never a signal. `abr.ts` has been formatting
//  fps, bitrate, decode latency, loss and RTT into `formatStatsLine()` since
//  the day the ABR controller was written; nothing downstream could plot any
//  of it, alert on any of it, or compare two stations on it, because it was
//  prose. This module takes the same tick and emits the numbers AS NUMBERS.
//
//  IT IS NOT AN EVENT LANE. Stalls, decode errors, ABR downshifts and blocked
//  audio are discrete and already have one (`analytics/streamEvents.ts`). What
//  goes here is the CONTINUOUS level such an event would be a threshold on —
//  which is why `framesDropped` rides as a cumulative counter and "the stream
//  stalled" is not in this file at all.
//
//  IT IS NOT `analytics/metrics.ts`. That plane is buckets, by explicit policy,
//  and its whole contract is aggregates with no per-sample timestamp. A time
//  series is a different shape, not a finer setting of the same one.
// ============================================================================

import { clientSessionId } from '../clientDebug';

/** One sample's numbers. Keys are the store's column names verbatim
 *  (`scripts/serve/vitals_schema.py` CATALOGUE) — deliberately not prettier
 *  ones, so that a value can be traced from this file to a chart axis without
 *  a translation table in between. A key the server does not know is ignored
 *  there rather than refused, so this may ship a vital before a deploy. */
export type VitalSample = Record<string, number>;

interface Envelope {
  station: string;
  sessionId: string;
  build: string;
}

/**
 * THE SAMPLE INTERVAL — one second, and it is chosen from what makes the
 * SIGNAL good, not from a capacity budget.
 *
 * The events this series has to make visible are short. A freeze latches after
 * 250 ms without decoded output; an ABR downshift and its recovery both happen
 * inside the controller's own 3-second rolling window; an audio underrun is
 * instantaneous. At the 5 s cadence the log line uses, a downshift and the
 * recovery from it can BOTH fall between two samples and the series shows a
 * flat line through a visible fault. At 1 s every ABR window contains three
 * samples and no freeze episode can hide entirely between two.
 *
 * ONE SECOND IS ALSO THE NATURAL FLOOR, which is why it is not faster. Four of
 * the vitals in every row are the DAEMON's own view, and `transport/mod.rs`
 * emits those at exactly 1 Hz. Sampling faster would repeat half of each row
 * verbatim — more rows, no more information.
 */
const SAMPLE_MS = 1_000;
/** How often the queue is drained to the server. NOT the sample interval:
 *  20 samples ride one POST. Twenty seconds matches the analytics sink's
 *  cadence, which is the interval this codebase has already decided a
 *  background POST may cost a visitor. */
const FLUSH_MS = 20_000;
/** Queue ceiling. At 1 s that is ~10 minutes of a backgrounded tab, far beyond
 *  what any flush gap can legitimately accumulate; past it the OLDEST are
 *  dropped, because in a stream fault the newest samples describe the fault. */
const MAX_PENDING = 600;
/** Server-side body cap is 512 KiB (`vitals.BODY_MAX`); stop well under it. */
const MAX_BATCH_CHARS = 200_000;

let pending: Array<{ t: number; v: VitalSample }> = [];
let envelope: Envelope | null = null;
let lastSampleAt = 0;
let timer: ReturnType<typeof setTimeout> | null = null;
let hooked = false;

/** u32 microsecond difference, as a SIGNED millisecond number.
 *
 *  Both media stamp their capture time with the server's µs clock truncated to
 *  32 bits, so the counter wraps every 2^32 µs ≈ 71.6 minutes and a naive
 *  subtraction across a wrap reports ±71 minutes of skew. Interpreting the
 *  difference as a signed 32-bit quantity gives the right answer for any true
 *  skew under ±35.8 minutes, which is every skew that is not already a bug of
 *  a different kind. Exported for the test. */
export function skewMs(audioTsUs: number, videoTsUs: number): number {
  const d = ((audioTsUs - videoTsUs) | 0); // two's-complement wrap, in µs
  return d / 1000;
}

/**
 * Name the producer. Called once per stream session, before the first sample.
 *
 * `station` becomes the store's `station` column and, at the far end of the
 * pipeline, the OTLP `service.instance.id` — which is the single field that
 * makes each exhibit its OWN monitored entity rather than a label on one
 * blurred service. It is worth getting right: an empty station id here is a
 * sample that lands under "unknown" and merges with every other one.
 */
export function beginVitals(station: string, build: string): void {
  envelope = { station: station || 'unknown', sessionId: clientSessionId(), build: build || 'unknown' };
  lastSampleAt = 0;
}

/**
 * Is a sample due? The cadence lives HERE and not in `abr.ts` because it is a
 * property of the vitals lane, not of the ABR tick that happens to drive it —
 * the tick runs at ~100 ms and its own log line at 5 s, and neither of those
 * numbers should move because this one does. The sink is single-session by
 * construction (one `envelope`), so one module-level clock is the right shape;
 * `beginVitals` resets it so a new station starts sampling immediately.
 */
export function dueForVitals(now: number): boolean {
  if (!envelope || now - lastSampleAt < SAMPLE_MS) return false;
  lastSampleAt = now;
  return true;
}

/** Queue one sample. Cheap and unconditional; the caller has already been told
 *  by `dueForVitals` that one is owed, and must not be given a second reason
 *  to think about cost. */
export function recordVitals(v: VitalSample): void {
  if (!envelope) return;
  try {
    // A sample with nothing finite in it is not a sample. Filtering here and
    // not at the server keeps a dead client from spending a POST on nulls.
    const clean: VitalSample = {};
    for (const [k, n] of Object.entries(v)) {
      if (typeof n === 'number' && Number.isFinite(n)) clean[k] = Math.round(n * 1000) / 1000;
    }
    if (!Object.keys(clean).length) return;
    pending.push({ t: Date.now(), v: clean });
    if (pending.length > MAX_PENDING) pending.splice(0, pending.length - MAX_PENDING);
    ensureTimer();
    ensurePagehideFlush();
  } catch { /* telemetry must never break the app */ }
}

function ensureTimer(): void {
  if (timer != null || typeof setTimeout === 'undefined') return;
  timer = setTimeout(() => { timer = null; void flushVitals(); }, FLUSH_MS);
}

function ensurePagehideFlush(): void {
  if (hooked || typeof window === 'undefined') return;
  hooked = true;
  // THE SAMPLES THAT MATTER MOST ARE THE LAST ONES. A stream that degrades
  // until the visitor gives up and closes the tab takes its final minute with
  // it unless this fires — which is exactly the minute an investigation wants.
  const go = () => { void flushVitals(true); };
  window.addEventListener('pagehide', go);
  window.addEventListener('visibilitychange', () => { if (document.visibilityState === 'hidden') go(); });
}

/**
 * Ship the queue. Failures FOLD BACK into the queue rather than being dropped:
 * a stream fault and a network fault arrive together often enough that
 * discarding on error would systematically lose the worst sessions — the
 * survivorship bias that makes a health dashboard say everything is fine.
 * An HTTP REFUSAL (4xx) does drop, because re-offering a body the server has
 * already judged malformed is a loop.
 */
export async function flushVitals(final = false): Promise<void> {
  if (!envelope || !pending.length || typeof fetch === 'undefined') return;
  const batch = pending;
  pending = [];
  const body = JSON.stringify({
    resource: {
      'service.instance.id': envelope.station,
      'session.id': envelope.sessionId,
      'kh.bundle': envelope.build,
      'kh.source': 'spa',
    },
    samples: batch,
  });
  if (body.length > MAX_BATCH_CHARS) {
    // Over the wire budget. Keep the NEWEST half and drop the oldest — the
    // opposite of what a queue usually does, and right here for the same
    // reason MAX_PENDING drops from the front: in a stream fault the recent
    // samples describe the fault and the old ones describe the calm before
    // it. The kept half goes back for the next tick rather than being sent
    // now, so this path costs a flush interval and never a request that the
    // server would refuse.
    pending = batch.slice(Math.floor(batch.length / 2)).concat(pending).slice(-MAX_PENDING);
    return;
  }
  try {
    const res = await fetch('/vitals', {
      method: 'POST',
      // `keepalive` is what makes the pagehide flush actually leave the tab.
      keepalive: final,
      credentials: 'same-origin',
      headers: { 'Content-Type': 'application/json' },
      body,
    });
    if (!res.ok && res.status >= 500) pending = batch.concat(pending).slice(-MAX_PENDING);
  } catch {
    pending = batch.concat(pending).slice(-MAX_PENDING);
  }
}

/** Drop everything and forget the producer. Called when a stream session ends
 *  so the next one does not inherit the previous station's envelope. */
export function endVitals(): void {
  void flushVitals(true);
  envelope = null;
  if (timer != null) { clearTimeout(timer); timer = null; }
}
